src/core/src/op/unsqueeze.cpp

```diff
@@ -82,29 +62,30 @@ bool evaluate(const HostTensorPtr& arg0, const HostTensorPtr& out) {
     return true;
 }
 
-bool evaluate_unsqueeze(const HostTensorPtr& arg0, const HostTensorPtr& arg1, const HostTensorPtr& out) {
+bool evaluate_unsqueeze(const Node* node,
+                        const HostTensorPtr& arg0,
+                        const HostTensorPtr& arg1,
+                        const HostTensorPtr& out) {
     auto element_type = arg0->get_element_type();
     out->set_element_type(element_type);
 
-    auto data_shape = arg0->get_shape();
-    int64_t data_rank = static_cast<int64_t>(data_shape.size());
-    auto axes_shape = arg1->get_shape();
-    NGRAPH_CHECK(axes_shape.size() == 1 || axes_shape.empty(),
-                 "Axes to add must be a scalar or 1D tensor with 1 element");
+    const auto& axes_shape = arg1->get_shape();
+    OPENVINO_ASSERT(ov::is_scalar(axes_shape) || ov::is_vector(axes_shape),
+                    "Axes to add must be a scalar or 1D tensor with 1 element");
+
+    const auto& data_shape = arg0->get_shape();
+    const auto out_rank = static_cast<int64_t>(data_shape.size() + shape_size(axes_shape));
+
+    // Get axes and normalize
+    auto axes = read_index_vector(arg1);
+    normalize_axes(node, out_rank, axes);
 
-    auto out_shape = data_shape;
-    int64_t out_rank = data_rank + static_cast<int64_t>(shape_size(axes_shape));
-    // Get axes
-    vector<int64_t> axes = read_index_vector(arg1);
-    // Normalize axes
-    std::transform(axes.begin(), axes.end(), axes.begin(), [out_rank](int64_t i) -> int64_t {
-        return i < 0 ? out_rank + i : i;
-    });
     // Sort in increasing order
-    std::set<int64_t, less<int64_t>> axes_set(axes.begin(), axes.end());
+    std::set<int64_t> axes_set(std::make_move_iterator(axes.begin()), std::make_move_iterator(axes.end()));
     NGRAPH_CHECK(axes.size() == axes_set.size(), "Axes has duplicate axis.");
+
+    auto out_shape = data_shape;
     for (int64_t axis : axes_set) {
-        NGRAPH_CHECK(axis >= 0 && axis < out_rank, "Axis is out of bounds: ", axis);
         out_shape.insert(out_shape.begin() + axis, 1);
```

Comment:
Shape infer and evaluate are not calculating output shape in same way and also use different validation.
Maybe in next version of operator it can be unified if shape infer follow `numpy` and reject repeated axes, then this shape calculations should be same.